### PR TITLE
Fixes #RHINENG-14713 - Error if no valid workload_type found

### PR DESCRIPTION
**Pull request by upadhyeammit**
## PR Title :boom:

Please title this PR with a summary of the change, along with the JIRA card number.

Suggested formats: 

1. Fixes/Refs #RHIROS-XXX - Title
2. RHIROS-XXX Title
3. RHINENG-XXX Title 

Feel free to remove this section from PR description once done.

## Why do we need this change? :thought_balloon:

Please include the __context of this change__ here. If the change depends on Kruize add details about that as well!

## Documentation update? :memo:

- [ ] Yes
- [ ] No

## Security Checklist :lock:

Upon raising this PR please go through [RedHatInsights/secure-coding-checklist](https://github.com/RedHatInsights/secure-coding-checklist)

## :guardsman: Checklist :dart:

- [ ] Does this change depend on specific version of Kruize
  - If yes what is the version no:
  - [ ] Is that image available in production or needs deployment?
- [ ] Bugfix
- [ ] New Feature
- [ ] Refactor
- [ ] Unittests Added
- [ ] DRY code
- [ ] Dependency Added
- [ ] DB Migration Added

## Additional :mega:

Feel free to add any other relevant details such as __links, notes, screenshots__, here.

### CODE REVIEW

**Comment by upadhyeammit** [comment]
@saltgen as discussed I saw the potential to add this in `filter_valid_csv_records` however in this method we dont have the column `k8s_object_type` created and it happens in later in `determine_k8s_object_type`.

With the code change the processor does not panic and errors out,
~~~
ERRO[0005]/home/amit/work/insights/ros/ros-ocp-backend/internal/services/report_processor.go:80 github.com/redhatinsights/ros-ocp-backend/internal/services.ProcessReport() Error: no valid k8s_object_types present in CSV to process further  account= cluster_alias=name222 cluster_uuid=222 org_id=3340851 request_id=uuid1234 service=rosocp source_id=111
~~~

**Review by saltgen — CHANGES_REQUESTED** [review]
@upadhyeammit As discussed, let's add the filter for workload type, on filter_csv_records.
This enum equivalent should help you, /internal/types/workload.

File: internal/utils/aggregator.go [hunk 1]
```diff
@@ -86,6 +82,37 @@ func Aggregate_data(df dataframe.DataFrame) (dataframe.DataFrame, error) {
 }
 
 func filter_valid_csv_records(main_df dataframe.DataFrame) (dataframe.DataFrame, int) {
+
+	emptyWorkload := func() func(el series.Element) bool {
+		return func(el series.Element) bool {
+			if el.Type() == series.String {
+				if workload, ok := el.Val().(string); ok {
+					if workload == "<none>" || workload == "" {
```
Comment (upadhyeammit) [inline internal/utils/aggregator.go:90]: The main branch code accepts `""` and `"<none>" workloads. I like to know should we filter it out then?

File: internal/utils/aggregator.go [hunk 2]
```diff
@@ -104,17 +101,8 @@ func filter_valid_csv_records(main_df dataframe.DataFrame) (dataframe.DataFrame,
...
-	no_of_dropped_records := main_df.Nrow() - df.Nrow()
-
-	return df, no_of_dropped_records
-}
-
-func filter_valid_k8s_object_types(df dataframe.DataFrame) dataframe.DataFrame {
-	return df.Filter(
 		dataframe.F{
-			Colname:    "k8s_object_type",
+			Colname:    "workload_type",
```
Comment (patilsuraj767) [inline internal/utils/aggregator.go:105]: Any specific reason we changed the colname from `k8s_object_type` to `workload_type`? 
As far as I remember `workload_type` does not tell the correct representation of workload. For example - even if the workload is of type replicaset, `workload_type` field is set to `deployment`.  Due to this behaviour we have a function `determine_k8s_object_type` which determines the correct type of workload and sets it in `k8s_object_type` column.

**Review by patilsuraj767 — DISMISSED** [review]

File: internal/utils/aggregator.go (hunk 2 above)
Comment (upadhyeammit) [reply internal/utils/aggregator.go:105]: Actually let me explain more maybe,
The failure was happening because there were no rows which had valid `k8s_object_type` or I can say it was empty. However we did not had `df.Nrows==0` after filtering the valid `k8s_object_type` and so we ended up giving empty df to Groups further.
It looked we can just check `df.Nrows==0` after `filter_valid_k8s_object_types` and that should fix it, and it can. However then it also felt we can drop the empty and non valid rows in `filter_valid_csv_records` based on the workload_type, and then do the `determine_k8s_object_type` . So indeed we do not need to do `filter_valid_k8s_object_types` after `determine_k8s_object_type`?

Comment (patilsuraj767) [reply internal/utils/aggregator.go:105]: Got your point. I agree we can apply filter before calling `determine_k8s_object_type`. But my concern is that you have removed the `filter_valid_k8s_object_types` function which actually filters workloads based on there workload_type. So in this case workload of type `job` will also get pass.

Comment (patilsuraj767) [reply internal/utils/aggregator.go:105]: Instead of comparing actual workload type, can we simply add here `dataframe.F{Colname: "workload_type", Comparator: series.Neq, Comparando: ""},`

Comment (patilsuraj767) [reply internal/utils/aggregator.go:105]: and keep `filter_valid_k8s_object_types` function as it is.

File: .~lock.output.csv#
Comment (saltgen) [inline .~lock.output.csv#:1]: Let's remove this file

**Review by saltgen — CHANGES_REQUESTED** [review]
Some changes were reviewed and suggestions provided over DM.

- var name of workloadType index
- single column for workloadType lowercase

File: internal/utils/aggregator_test.go [hunk 3]
```diff
@@ -93,9 +120,8 @@ func Test_filter_valid_k8s_object_types(t *testing.T) {
 			"1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1",
 		},
 	}
-	df := dataframe.LoadStructs(usage_data)
-	df = determine_k8s_object_type(df)
-	result := filter_valid_k8s_object_types(df)
+	df = dataframe.LoadStructs(usage_data)
+	result, _ = filter_valid_csv_records(df)
 	fmt.Println(result.Nrow())
```
Comment (saltgen) [inline internal/utils/aggregator_test.go:125]: Let's remove this statement

File: internal/utils/aggregator_test.go
Comment (saltgen) [inline internal/utils/aggregator_test.go:1]: @upadhyeammit Since these changes cover the edge case being handled when workload and workload_name is empty, we should have a test case for the same, let's say after L#136?

**Review by saltgen — CHANGES_REQUESTED** [review]

File: internal/utils/aggregator.go [hunk 4]
```diff
@@ -104,27 +100,48 @@ func filter_valid_csv_records(main_df dataframe.DataFrame) (dataframe.DataFrame,
 		dataframe.F{Colname: "owner_name", Comparator: series.Neq, Comparando: ""},
 		dataframe.F{Colname: "owner_kind", Comparator: series.Neq, Comparando: "<none>"},
 		dataframe.F{Colname: "owner_name", Comparator: series.Neq, Comparando: "<none>"},
+		dataframe.F{Colname: "workload_type", Comparator: series.Neq, Comparando: "<none>"},
+		dataframe.F{Colname: "workload_type", Comparator: series.Neq, Comparando: ""},
 	)
 
-	no_of_dropped_records := main_df.Nrow() - df.Nrow()
+	// The above filters can delete all the rows
+	if df.Nrow() == 0 {
```
Comment (upadhyeammit) [inline internal/utils/aggregator.go:108]: @saltgen writing more tests unearth this situation, added a code for same!

File: internal/utils/aggregator_test.go [hunk 5]
```diff
@@ -178,3 +190,27 @@ func Test_check_if_all_required_columns_in_CSV(t *testing.T) {
 		t.Error("Expecting error to be returned as all required column not present")
 	}
 }
+
+func Test_Aggregate_data(t *testing.T) {
```
Comment (upadhyeammit) [inline internal/utils/aggregator_test.go:194]: Added a test on Aggregate to attest condition which fixed this bug!

File: internal/utils/aggregator_test.go [hunk 6]
```diff
@@ -112,43 +136,31 @@ func Test_filter_valid_k8s_object_types(t *testing.T) {
...
-func Test_filter_valid_csv_records(t *testing.T) {
-	usage_data := []UsageData{
-		// k8s object with missing data
-		{
-			"2023-02-01 00:00:00 +0000 UTC", "2023-03-01 00:00:00 +0000 UTC", "2023-06-02 00:00:01 +0000 UTC", "2023-06-02 00:15:00 +0000 UTC",
-			"Yuptoo-service", "Yuptoo-app-standalone-1", "Yuptoo-app", "ReplicaSet", "testdeployment", "deployment", "Yuptoo-prod",
-			"quay.io/cloudservices/yuptoo", "ip-10-0-176-227.us-east-2.compute.internal", "i-0dfbb3fa4d0e8fc94",
-			"1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "",
-		},
+	// check if empty workload_type is dropped
```
Comment (saltgen) [inline internal/utils/aggregator_test.go:144]: I don't think we'll need to check for the empty/`<none>` workloadType here, anymore.
This is because you have added the changes in `Test_Aggregate_data` that,

1. checks for empty, `<none>` workload_type
2. gracefully asserts that function does not panic

File: internal/utils/aggregator_test.go (hunk 5 above)
Comment (saltgen) [reply internal/utils/aggregator_test.go:194]:
```suggestion
func TestAggregateDataNoRecords(t *testing.T) {
```

**Review by saltgen — CHANGES_REQUESTED** [review]

File: internal/utils/aggregator_test.go (hunk 5 above)
Comment (saltgen) [reply internal/utils/aggregator_test.go:194]:
```suggestion
func TestAggregateDataNoRecords(t *testing.T) {
```

File: internal/utils/aggregator.go (hunk 4 above)
Comment (saltgen) [reply internal/utils/aggregator.go:108]: Good Stuff! 🌟

File: internal/utils/aggregator_test.go (hunk 6 above)
Comment (upadhyeammit) [reply internal/utils/aggregator_test.go:144]: hmm, we have tests for most of the filters? That does mean in future if someone removes these filters for dropping empty `workload_type` then the test should fail?
The Test_aggregate_data is using same `usage_data` however intention is different, and its about `gracefully asserts that function does not panic`

**Review by saltgen — APPROVED** [review]
Thanks for the changes @upadhyeammit. Appreciate your involvement and approach! 🙌🏼

**Comment by upadhyeammit** [comment]
> Thanks for the changes @upadhyeammit. Appreciate your involvement and approach! 🙌🏼

thanks for all the suggestions and help!
